### PR TITLE
Make Lazy Decomit asynchronous

### PR DESCRIPTION
**Pull request by mjp41**
On platforms that support low-memory notifications register callbacks
that perform lazy decommit. This allows idle processes to return memory
to the OS. Without incurring the cost of constantly committing and
decommitting memory.

### CODE REVIEW

**Comment by mjp41** [comment]
The addresses the discussion in #127.

@davidchisnall I am a bit unhappy with the unsafe downcast, but also don't want to pull in vtables and RTTI for downcasts.  So thoughts on how this could be better welcome.

File: src/pal/pal_windows.h [hunk 1]
```diff
@@ -20,23 +20,31 @@ namespace snmalloc
...
+
     /**
      * Callback, used when the system delivers a low-memory notification.  This
-     * simply increments an atomic counter each time the notification is raised.
+     * calls all the handlers registered with the PAL.
      */
     static void CALLBACK low_memory(_In_ PVOID, _In_ BOOLEAN)
     {
-      pressure_epoch++;
+      PalNotificationObject* curr = low_memory_callbacks;
```
Comment (davidchisnall) [inline src/pal/pal_windows.h]: Can this and the list be pulled out into a separate class?  I'd rather not have to copy and paste this for other platforms.

Comment (mjp41) [reply src/pal/pal_windows.h]: Done

**Comment by mjp41** [comment]
@buybackoff I have modified the test here, so you can run the new code in several parallel processes.  The test runs until it exhausts physical RAM, and then deallocates everything, and waits for a key press.  If you then stress the machine (say by running another copy of the test), then the RAM usages drops down to almost zero, on my machine 10GiB -> 34MiB, while it is sitting waiting for user input.  I think this matches perfectly the scenario you were seeing.

**Comment by buybackoff** [comment]
@mjp41 thanks! Will give it a try.

Should shared memory (SM) paged-in page cause lazy decommit? Or new SM paged-in pages will only evict other SM pages while snmalloc's pages are not yet decommited?

**Comment by mjp41** [comment]
So we are just performing lazy decommit on a low memory notification:

https://docs.microsoft.com/en-us/windows/win32/api/memoryapi/nf-memoryapi-creatememoryresourcenotification

This says it is when "available physical memory is running low".  'snmalloc' will return any "chunk" 1MiB or 16MiB range of memory we are no longer using.  Any allocations on the chunk will cause it to remain resident.  Decommiting memory at a finer granularity would involve much more synchronization on fairly common paths, so I am not currently keen to do this.

If the shared-memory is using physical pages then I believe it can trigger 'snmalloc' to give up memory it does not need.
